src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Primitives
+{
+    public struct StringSegment : IEquatable<StringSegment>
+    {
+        private readonly string _buffer;
+        private readonly int _offset;
+        private readonly int _count;
+
+        public StringSegment(string buffer, int offset, int count)
+        {
+            _buffer = buffer;
+            _offset = offset;
+            _count = count;
+        }
+
+        public string Buffer
+        {
+            get { return _buffer; }
+        }
+
+        public int Offset
+        {
+            get { return _offset; }
+        }
+
+        public int Count
+        {
+            get { return _count; }
+        }
+
+        public string Value
+        {
+            get { return _offset == -1 ? null : _buffer.Substring(_offset, _count); }
```

Comment:
These are fragile, and need to handle the case where the default constructor is used.